numpy/core/src/multiarray/multiarraymodule.c

```diff
@@ -2369,6 +2471,50 @@ array_vdot(PyObject *NPY_UNUSED(dummy), PyObject *args)
 }
 
 
+NPY_NO_EXPORT PyObject *
+array_matmul(PyObject *NPY_UNUSED(m), PyObject *args, PyObject* kwds)
+{
+    static PyUFuncObject *cached_npy_matmul = NULL;
+    int errval;
+    PyObject *override = NULL;
+    PyObject *v, *a, *o = NULL;
+    char* kwlist[] = {"a", "b", "out", NULL };
+
+    if (cached_npy_matmul == NULL) {
+        PyObject *module, *dict, *matmul;
+
+        module = PyImport_ImportModule("numpy.core.multiarray");
+        dict = PyModule_GetDict(module);
+        matmul = PyDict_GetItemString(dict, "matmul");
+        cached_npy_matmul = (PyUFuncObject*)matmul;
+        Py_INCREF(cached_npy_matmul);
+        Py_DECREF(module);
+    }
+
+    errval = PyUFunc_CheckOverride(cached_npy_matmul, "__call__",
+                                   args, kwds, &override, 2);
```

Comment:
I was also feeling a little hesitant about adding it to the C api, so +1 on
waiting until we're more sure what we're doing here.
On May 14, 2015 11:32 PM, "Charles Harris" <notifications@github.com> wrote:

> In numpy/core/src/multiarray/multiarraymodule.c
> <https://github.com/numpy/numpy/pull/5878#discussion_r30387829>:
>
> > +    PyObject *v, *a, *o = NULL;
> > +    char* kwlist[] = {"a", "b", "out", NULL };
> > +
> > +    if (cached_npy_matmul == NULL) {
> > +        PyObject *module, *dict, *matmul;
> > +
> > +        module = PyImport_ImportModule("numpy.core.multiarray");
> > +        dict = PyModule_GetDict(module);
> > +        matmul = PyDict_GetItemString(dict, "matmul");
> > +        cached_npy_matmul = (PyUFuncObject*)matmul;
> > +        Py_INCREF(cached_npy_matmul);
> > +        Py_DECREF(module);
> > +    }
> > +
> > +    errval = PyUFunc_CheckOverride(cached_npy_matmul, "__call__",
> > +                                   args, kwds, &override, 2);
>
> That raises the question if it should be in the C-API at this point. I'm
> tempted to leave it out and just go with the matmul method.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/numpy/numpy/pull/5878/files#r30387829>.
>